Tensile/SolutionStructs.py

```diff
@@ -1801,12 +1801,46 @@ def initBetaOnlyKernelObjects(self):
   # create Conversion Kernels
   def initConversionKernelObjects(self):
     self.conversionKernelObjects = []
-    if (self["GlobalSplitU"] > 1) and self["_GlobalAccumulation"]:
-      state = {}
-      state["ProblemType"] = deepcopy(self["ProblemType"])
-      state["KernelLanguage"] = "Source"
-      state["_GlobalAccumulation"] = self["_GlobalAccumulation"]
-      self.conversionKernelObjects.append(KernelWriterConversion(state))
+    gsu = self["GlobalSplitU"]
+    if (gsu > 1) and self["_GlobalAccumulation"]:
+      # wider load for GSU is single compute type only
+      supportedTypeForVWopt = self["ProblemType"]["ComputeDataType"].isSingle() or self["ProblemType"]["ComputeDataType"].isDouble()
+      vwMax = 1
+      if (supportedTypeForVWopt):
+        vwMax = 2
+
+      # reduction for GSU is single compute type + gus = power of 2 only
+      supportedTypeForReductionOpt = self["ProblemType"]["ComputeDataType"].isSingle() or self["ProblemType"]["ComputeDataType"].isDouble()
+      maxReduction = 1
+      maxReductionConst = 4 # this must match the value in client code (ContractionSolution.cpp)
+      minGSUperReduction = 32; # Minimum GSU=128 for Reduction=4, GSU=64 for Reduction2
+      applicableReduction = max(1, gsu // minGSUperReduction)
+      if (supportedTypeForReductionOpt and ((gsu & (gsu - 1)) == 0) and self["_GlobalAccumulation"] == "MultipleBuffer"):
+        maxReduction = min(applicableReduction, maxReductionConst) # not exceeding reductionThreshold
+
+      # loop unroll opt for postGSU
+      supportedTypeForUnrollOpt = self["ProblemType"]["ComputeDataType"].isSingle() or self["ProblemType"]["ComputeDataType"].isDouble()
+
+      vw = 1
+      while vw <= vwMax:
+        reduction = 1
+        while reduction <= maxReduction:
+          # so far, reduction=2 does not perform well. Skip 2
+          if reduction == 2 and self["ProblemType"]["ComputeDataType"].isSingle():
+            reduction *= 2
+            continue
+          state = {}
+          state["ProblemType"] = deepcopy(self["ProblemType"])
+          state["KernelLanguage"] = "Source"
+          state["_GlobalAccumulation"] = self["_GlobalAccumulation"]
+          state["GlobalSplitU"] = self["GlobalSplitU"]
+          state["VectorWidth"] = vw
+          state["Reduction"] = reduction
+          # number of unroll for large GSU (must match client code)
+          state["GSUUnrollUnit"] = 16 * state["Reduction"] if supportedTypeForUnrollOpt and self["_GlobalAccumulation"] == "MultipleBuffer" else 1
+          self.conversionKernelObjects.append(KernelWriterConversion(state))
```

Comment:
I wonder what the performance impact would be if some of these were run-time parameters to reduce the number of kernel variations generated

Reply:
> I wonder what the performance impact would be if some of these were run-time parameters to reduce the number of kernel variations generated

In that case, we need to add if statements in postGSU kernel.
There are 3 parameters to generate different postGSU kernels
   vector width
   reduction
   gsu (unroll)
GSU value is related to loop unroll.
It is not just about if statement.
Adding these if statements is probably not good for performance.
I already limited number of maximum unroll and used same postGSU kernels for larger GSU values.

The number of maximum postGSU kernels is
   GSU: 31 (from 2 to 32)
   Reduction: 2 for F32, 3 for F64
   VW: 2

Total is 124 for F32, 186 for F64.
(this is in case we generate kernels with all GSU values (from 2 to 32)

Compare to main asm kernels, the size of postGSU kernels is much smaller.
I do not have confirmed actual code sizes, but I am guessing the amount of postGSU kernels is probably equivalent to several main kernels.

To reduce the amount of kernels, we can add gsu parameter to main (asm) kernels.
(same idea as hipBlasLt)